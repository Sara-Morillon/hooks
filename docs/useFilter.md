# useFilter

`useFilter` offers an easy way to apply local filters to a set of local data.

## Example

```tsx
import { useFilter } from '@saramorillon/hooks'

interface IData {
  name: string
  age: number
}

const data: IData[] = [
  { name: 'Zola Ray', age: 30 },
  { name: 'Arlo Curtis', age: 32 },
]

function MyComponent() {
  const { rows, state, filter } = useFilter<IData>(data)

  return (
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Age</th>
        </tr>

        <tr>
          <td>
            <input value={state.name} onChange={(e) => filter('name', e.target.value)} />
          </td>
          <td>
            <input type="number" value={state.age} onChange={(e) => filter('age', e.target.value)} />
          </td>
        </tr>
      </thead>

      <tbody>
        {rows.map((row) => (
          <tr>
            <td>{row.name}</td>
            <td>{row.age}</td>
          </tr>
        ))}
      </tbody>
    </table>
  )
}
```

## API

```typescript
useFilter<T, F extends IFilter<T>>(data: T[], initialFilter?: Partial<F>, filterFunctions?: IFilterFunctions<T, F>): IFiltered<T, F>
```

### Arguments

`data: T[]` - the data that the state will by applied to.

`initialFilter?: Partial<F>` - the initial filters, default to an empty object.

`filterFunctions: IFilterFunctions<T, F>` - the functions to use to filter the data.

### Returns

`IFiltered<T, F>` - an object containing:

- `rows: T[]` - the filtered rows.
- `state: Partial<F>` - the filter state.
- `filter: <K extends keyof T>(field: K, value: F[K]) => void` - a function to apply a filter to a field.
