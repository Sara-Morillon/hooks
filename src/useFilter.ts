import { useCallback, useMemo, useState } from 'react'

export type IFilter<T> = {
  [key in keyof T]: unknown
}

export type IFilterFunctions<T, F extends IFilter<T> = T> = {
  [column in keyof T]?: (row: T, filter: F[column]) => boolean
}

export function useFilter<T, F extends IFilter<T> = T>(data: T[], filterFunctions?: IFilterFunctions<T, F>) {
  const [state, setState] = useState<Partial<F>>({})

  const filter = useCallback(<K extends keyof F>(column: K, value?: F[K]) => {
    setState((state) => ({ ...state, [column]: value }))
  }, [])

  const rows = useMemo(() => {
    return data.filter((row) => {
      return Object.entries(state).every((entry) => {
        const column = entry[0] as keyof T
        const value = entry[1] as F[keyof T]
        if (value === undefined || value === null || value === '' || (Array.isArray(value) && !value.length)) {
          return true
        }
        const filterFunction = filterFunctions?.[column]
        if (filterFunction) {
          return filterFunction(row, value)
        }
        return row[column] === value
      })
    })
  }, [data, state, filterFunctions])

  return useMemo(() => ({ state, filter, rows }), [state, filter, rows])
}
