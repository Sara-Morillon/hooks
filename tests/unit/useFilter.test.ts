import { act, renderHook } from '@testing-library/react'
import { type IFilterFunctions, useFilterState, useFilteredRows } from '../../src/useFilter.js'
import { type IData, mockTableData } from '../mock.js'

describe('useFilterState', () => {
  it('should return empty filters by default', () => {
    const { result } = renderHook(() => useFilterState<IData>())
    expect(result.current.state).toEqual({})
  })

  it('should change filter', () => {
    const { result } = renderHook(() => useFilterState<IData>())
    act(() => result.current.filter('name', 'Zola Ray'))
    expect(result.current.state).toEqual({ name: 'Zola Ray' })
  })

  it('should override existing filter', () => {
    const { result } = renderHook(() => useFilterState<IData>())
    act(() => result.current.filter('name', 'Zola Ray'))
    act(() => result.current.filter('name', 'Jaxen Smith'))
    expect(result.current.state).toEqual({ name: 'Jaxen Smith' })
  })
})

describe('useFilteredRows', () => {
  it('should return all rows if filter is empty', () => {
    const { result } = renderHook(() => useFilteredRows(mockTableData()))
    expect(result.current).toEqual(mockTableData())
  })

  it('should not filter rows if filter is an empty string', () => {
    const { result } = renderHook(() => useFilteredRows(mockTableData(), { name: '' }))
    expect(result.current).toEqual(mockTableData())
  })

  it('should not filter rows if filter is an empty ', () => {
    const { result } = renderHook(() => useFilteredRows(mockTableData(), { name: [] }))
    expect(result.current).toEqual(mockTableData())
  })

  it('should filter rows with default filter function', () => {
    const { result } = renderHook(() => useFilteredRows(mockTableData(), { name: 'Zola Ray' }))
    expect(result.current).toEqual([{ name: 'Zola Ray', age: 30 }])
  })

  it('should filter rows with custom filter function', () => {
    const filterFunctions: IFilterFunctions<IData> = { name: (row, value) => row.name.includes(value) }
    const { result } = renderHook(() => useFilteredRows(mockTableData(), { name: 'x' }, filterFunctions))
    expect(result.current).toEqual([
      { name: 'Alexis West', age: 39 },
      { name: 'Jaxen Smith', age: 35 },
    ])
  })
})
