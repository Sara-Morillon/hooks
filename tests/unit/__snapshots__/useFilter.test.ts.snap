// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useFilter > should filter rows with custom filter function 1`] = `
[
  {
    "age": 39,
    "name": "Alexis West",
  },
  {
    "age": 35,
    "name": "Jaxen Smith",
  },
]
`;

exports[`useFilter > should filter rows with default filter function 1`] = `
[
  {
    "age": 30,
    "name": "Zola Ray",
  },
]
`;

exports[`useFilteredRows > should filter rows with custom filter function 1`] = `
[
  {
    "age": 39,
    "name": "Alexis West",
  },
  {
    "age": 35,
    "name": "Jaxen Smith",
  },
]
`;

exports[`useFilteredRows > should filter rows with default filter function 1`] = `
[
  {
    "age": 30,
    "name": "Zola Ray",
  },
]
`;
